internal/wrphandlers/qos/priority_queue.go

```diff
@@ -0,0 +1,95 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"container/heap"
+	"sync"
+
+	"github.com/xmidt-org/wrp-go/v3"
+)
+
+// PriorityQueue implements heap.Interface and holds wrp Message, using wrp.QOSValue as its priority.
+// https://xmidt.io/docs/wrp/basics/#qos-description-qos
+type PriorityQueue struct {
+	queue []wrp.Message
+	// maxQueueSize is the allowable max size of the queue based on the sum of all queued wrp message's payloads
+	maxQueueSize int
+
+	m sync.Mutex
+}
+
+// Dequeue returns the next highest priority message.
+func (pq *PriorityQueue) Dequeue() (wrp.Message, bool) {
+	defer pq.m.Unlock()
+	pq.m.Lock()
+	if pq.Len() == 0 {
+		return wrp.Message{}, false
+	}
+
+	msg, ok := heap.Pop(pq).(wrp.Message)
```

Comment:
I'd not assign these, but just return from the `Pop()`

Reply:
yeah I wanted to do that but go doesn't let us do that and keep the boolean indicator of whether the assertion succeeded

<img width="615" alt="image" src="https://github.com/xmidt-org/xmidt-agent/assets/12867159/63c65065-3f4f-4ad6-874e-f245f2eb5d6f">